File: python/walayer/http.py
"""HTTP transport for the WALayer client.

The transport is injectable so tests (and alternative HTTP stacks) can swap it
without touching client logic — the same pattern as the Node SDK's injectable
``fetch``. The default uses the standard library only, so the package has zero
runtime dependencies.
"""
from __future__ import annotations

import json
from urllib.parse import urlencode
import urllib.error
import urllib.request
from typing import Any, Callable, Dict, Optional, Tuple

from .errors import WALayerError

# (method, url, headers, body_bytes) -> (status, response_headers, text)
Transport = Callable[[str, str, Dict[str, str], Optional[bytes]], Tuple[int, Dict[str, str], str]]


def default_transport(
    method: str, url: str, headers: Dict[str, str], body: Optional[bytes]
) -> Tuple[int, Dict[str, str], str]:
    req = urllib.request.Request(url, data=body, method=method, headers=headers)
    try:
        with urllib.request.urlopen(req) as resp:  # noqa: S310 (trusted base_url)
            return resp.status, {k.lower(): v for k, v in resp.headers.items()}, resp.read().decode("utf-8")
    except urllib.error.HTTPError as exc:
        text = exc.read().decode("utf-8") if exc.fp else ""
        return exc.code, {k.lower(): v for k, v in exc.headers.items()}, text


class Http:
    """Thin JSON transport: adds the Bearer key, unwraps ``{data}``, and raises a
    typed :class:`WALayerError` on any non-2xx response."""

    def __init__(self, api_key: str, base_url: str = "https://api.walayer.com", transport: Optional[Transport] = None):
        if not api_key:
            raise ValueError("WALayer: api_key is required")
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._transport = transport or default_transport

    @staticmethod
    def _qs(params: Optional[Dict[str, Any]]) -> str:
        """Build ?a=1&b=2, dropping unset values so callers can pass kwargs through."""
        if not params:
            return ""
        pairs = {k: v for k, v in params.items() if v is not None and v != ""}
        if not pairs:
            return ""
        return "?" + urlencode({k: ("true" if v is True else "false" if v is False else v)
                                for k, v in pairs.items()})

    def request(
        self,
        path: str,
        method: str = "GET",
        body: Any = None,
        extra_headers: Optional[Dict[str, str]] = None,
        params: Optional[Dict[str, Any]] = None,
        envelope: bool = False,
    ) -> Any:
        """
        `envelope=True` returns the whole response body rather than just `data`.

        List endpoints answer `{"data": [...], "next_cursor": ...}`; unwrapping to
        `data` silently discarded the cursor, so a caller could read the first
        page and had no way to ask for the second.
        """
        headers = {"authorization": f"Bearer {self._api_key}"}
        if extra_headers:
            headers.update(extra_headers)
        data: Optional[bytes] = None
        if body is not None:
            headers["content-type"] = "application/json"
            data = json.dumps(body).encode("utf-8")

        url = f"{self._base_url}{path}{self._qs(params)}"
        status, resp_headers, text = self._transport(method, url, headers, data)
        request_id = resp_headers.get("x-request-id")

        if status == 204:
            return None
        parsed = json.loads(text) if text else {}
        if status >= 400:
            err = parsed.get("error", {}) if isinstance(parsed, dict) else {}
            raise WALayerError(status, err.get("code", "UNKNOWN"), err.get("message", ""), err.get("detail"), request_id)
        if envelope:
            return parsed
        return parsed.get("data") if isinstance(parsed, dict) else parsed

File: node/src/http.ts
import { WALayerError } from "./errors.js";

export interface HttpOptions {
  apiKey: string;
  baseUrl?: string;
  /** injectable for tests; defaults to global fetch */
  fetch?: typeof fetch;
}

interface RequestInitEx {
  method?: string;
  body?: unknown;
  headers?: Record<string, string>;
  /** appended as a query string; undefined and null values are dropped */
  query?: Record<string, string | number | boolean | undefined | null>;
}

/** A list response that carries paging state alongside the rows. */
export interface Page<T> {
  data: T[];
  next_cursor: string | null;
}

/**
 * Thin JSON transport over the WALayer REST API. Adds the Bearer key, unwraps
 * the `{ data }` envelope, and turns any non-2xx into a typed WALayerError with
 * the server's error code and request id — so callers `catch (e) { e.code }`
 * rather than parsing bodies.
 */
export class Http {
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly doFetch: typeof fetch;

  constructor(opts: HttpOptions) {
    if (!opts.apiKey) throw new Error("WALayer: apiKey is required");
    this.apiKey = opts.apiKey;
    this.baseUrl = (opts.baseUrl ?? "https://api.walayer.com").replace(/\/$/, "");
    this.doFetch = opts.fetch ?? globalThis.fetch;
    if (!this.doFetch) throw new Error("WALayer: no fetch available; pass one via options.fetch");
  }

  /** Build `?a=1&b=2`, dropping anything unset so callers can spread options. */
  private qs(query: RequestInitEx["query"]): string {
    if (!query) return "";
    const p = new URLSearchParams();
    for (const [k, v] of Object.entries(query)) {
      if (v !== undefined && v !== null && v !== "") p.set(k, String(v));
    }
    const s = p.toString();
    return s ? `?${s}` : "";
  }

  /**
   * The full response envelope, not just `data`.
   *
   * List endpoints return `{ data, next_cursor }` and `/v1/search` returns
   * `{ data, meta }`. `request()` unwraps to `data`, which silently threw the
   * cursor away — so an SDK user could read the first page and had no way to
   * ask for the second.
   */
  async envelope<T>(path: string, init: RequestInitEx = {}): Promise<T> {
    return this.send<T>(path, init, false);
  }

  async request<T>(path: string, init: RequestInitEx = {}): Promise<T> {
    return this.send<T>(path, init, true);
  }

  private async send<T>(path: string, init: RequestInitEx, unwrap: boolean): Promise<T> {
    const headers: Record<string, string> = {
      authorization: `Bearer ${this.apiKey}`,
      ...(init.headers ?? {}),
    };
    if (init.body !== undefined) headers["content-type"] = "application/json";

    const reqInit: RequestInit = { method: init.method ?? "GET", headers };
    if (init.body !== undefined) reqInit.body = JSON.stringify(init.body);
    const res = await this.doFetch(`${this.baseUrl}${path}${this.qs(init.query)}`, reqInit);

    const requestId = res.headers.get("x-request-id");
    if (res.status === 204) return undefined as T;

    const text = await res.text();
    const json = text ? (JSON.parse(text) as unknown) : {};

    if (!res.ok) {
      const err = (json as { error?: { code?: string; message?: string; detail?: unknown } }).error ?? {};
      throw new WALayerError(res.status, err.code ?? "UNKNOWN", err.message ?? res.statusText, err.detail, requestId);
    }
    return unwrap ? (json as { data: T }).data : (json as T);
  }
}
